### PR TITLE
feat(querying): Person on events querying

### PR DESCRIPTION
**Pull request by EDsCODE**
## Problem

- initial changes to use new person_properties
- [ ] Account for new and old events that may or may not have new fields populated
- [ ] Handle group querying
- [ ] Handle tests. All tests now need person_id when creating events for the queries to be easily testable
- [ ] Handle trends breakdown event query separatel (it doesn't use the base EventQuery class)

## Changes

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: posthog/models/utils.py [hunk 1]
```diff
@@ -22,6 +22,7 @@ class PersonPropertiesMode(Enum):
     USING_PERSON_PROPERTIES_COLUMN = auto()
     # Used for generating query on Person table
     DIRECT = auto()
+    USING_PERSON_ON_EVENT_COLUMNS = auto()
```
Comment (EDsCODE) [inline posthog/models/utils.py:25]: Might be able to just use DIRECT here instead of another field

File: ee/clickhouse/models/property.py [hunk 2]
```diff
@@ -225,6 +236,17 @@ def parse_prop_clauses(
...
+                idx,
+                prepend,
+                prop_var=f"$group{prop.group_type_index}_properties",
+                allow_denormalized_props=False,
+                property_operator=property_operator,
+            )
+            final.append(filter_query)
+            params.update(filter_params)
         elif prop.type == "group":
             if group_properties_joined:
```
Comment (EDsCODE) [inline ee/clickhouse/models/property.py:251]: Update that this isn't always true anymore

**Comment by marcushyett-ph** [comment]
Not sure if this is the best place to leave this comment - so please direct me somewhere else if here isn't:

We need to ensure we handle person property filtering and aggregation consistently when updating queries, today we essentially always use the "latest version of person property". To make the transition smooth, we should consider maintaining this behavior initially.

But I think we should aim to very quickly follow-up supporting a full range of options e.g. "initial properties", "properties on a specific date", "previous properties", "latest properties", "properties at time of event"

I believe the default should be "properties at time of event" going forward - to give us the benefits of performance and accuracy.

**Comment by neilkakkar** [comment]
Right now, I'm pretty opposed to going this route for a couple of reasons:

1. We're trying to adapt existing code with its assumptions to a completely new paradigm, which will make things very brittle and awkward. For example, there shall only be one `PersonPropertiesMode` at the end of this, `DIRECT` and `USING_SUBQUERY` will make zero sense, and `USING_PERSON_PROPERTIES_COLUMNS` is exactly what we'll be doing everywhere.
2. This makes maintaining existing code harder, makes bugs harder to think through, and can also introduce new subtle bugs in old querying if we make mistakes here.
3. As a function of the above two, this makes the final transition harder as well, since there's now old and new stuff mixed, and final refactors would need to clean all of this up.

So, I propose doing this a bit differently:

I think it's a fair assumption to make that we'll backfill all existing data to have person on events, etc. etc. Given this, we can completely get rid of the constraint that we need to support old and new style querying at the same time. Thus:

1. We have an instance setting which determines which type of querying to use
2. We rewrite the `*event_query` classes.  Yes, this means copying & pasting some stuff over, but we don't have to support any old paradigms, which will massively simplify things. The invariant though, is that to every trend/funnel/insight querying, nothing changes. The event query used to push up person properties, person_id, and event properties, and we continue doing exactly this, without the joins, and use mode `USING_PERSON_PROPERTIES_COLUMN` (which will be now materialised on events)
3. We can gradually flip this switch on for teams, and see everything still works, etc.

The thing here we might struggle with is `PERSON_TABLE_ALIASES`, but that should be straightforward enough to resolve.

Am I missing something that makes going this route much harder?

**Comment by neilkakkar** [comment]
@marcushyett-ph talk to me a bit more about this:

>  To make the transition smooth, we should consider maintaining this behavior initially.

Do we need to do this in code? Would talking to users / being clear about the changes coming make more sense?

I'm not sure if we talked about this before, but supporting these makes us lose the perf gains that were supposed to come with this 🤔 .

**Comment by marcushyett-ph** [comment]
> Do we need to do this in code? Would talking to users / being clear about the changes coming make more sense?

My concern is during the transition period (e.g. whilst we're backfilling events and writing new queries), we shouldn't suddenly change the results for some queries and not others - without informing people.

We 100% should be talking to users about the changes coming, explaining motivations, affects, timescale etc.

Will come up with a plan on how we inform / announce this with @joethreepwood when I'm back next week.

**Comment by EDsCODE** [comment]
> We have an instance setting which determines which type of querying to use

Can't go this route. We need granular control where you use the new query pattern if the query is bounded to a specific range where we know the data is already being ingested with the new format/backfilled

> For example, there shall only be one PersonPropertiesMode at the end of this, DIRECT and USING_SUBQUERY will make zero sense, and USING_PERSON_PROPERTIES_COLUMNS is exactly what we'll be doing everywhere.

Agreed, that's why I made a note that the new PersonPropertiesMode option might not be necessary

> We rewrite the *event_query classes. Yes, this means copying & pasting some stuff over, but we don't have to support any old paradigms, which will massively simplify things

What warrants this? We already have code paths that structurally are "querying directly" so if we take advantage of those for the time being to gradually roll out the new queries, once all the data is backfilled, we can prune all the unnecessary code paths and we end up with what we want anyway. 

Our eventquery logic right now is fairly clear in allowing you to decide how/when to join distinct_ids/person_ids and where the filtering is done. If anything, Marcus's point about time based properties is what would warrant an entire eventquery rewrite

**Comment by neilkakkar** [comment]
Had a chat, and sounds like I was mistaken, most of the complexity is shielded elsewhere, which makes refactoring much easier than I thought, so going forward this route for now!

File: ee/clickhouse/test/test_journeys.py [hunk 3]
```diff
@@ -68,12 +69,12 @@ def _create_all_events(all_events: List[Dict]):
...
         in_memory_event = InMemoryEvent(**data)
         parsed += f"""
-        ('{str(uuid4())}', '{in_memory_event.event}', '{json.dumps(in_memory_event.properties)}', '{in_memory_event.timestamp}', {in_memory_event.team.pk}, '{in_memory_event.distinct_id}', '', '{timezone.now().strftime("%Y-%m-%d %H:%M:%S.%f")}', now(), 0)
+        ('{str(uuid4())}', '{in_memory_event.event}', '{json.dumps(in_memory_event.properties)}', '{in_memory_event.timestamp}', {in_memory_event.team.pk}, '{in_memory_event.distinct_id}', '', '{in_memory_event.person_id}', '{timezone.now().strftime("%Y-%m-%d %H:%M:%S.%f")}', now(), 0)
         """
 
     sync_execute(
         f"""
-    INSERT INTO {EVENTS_DATA_TABLE()} (uuid, event, properties, timestamp, team_id, distinct_id, elements_chain, created_at, _timestamp, _offset) VALUES
+    INSERT INTO {EVENTS_DATA_TABLE()} (uuid, event, properties, timestamp, team_id, distinct_id, elements_chain, person_id, created_at, _timestamp, _offset) VALUES
```
Comment (EDsCODE) [inline ee/clickhouse/test/test_journeys.py:77]: Refactoring `journeys_for` is no problem but if we want to properly test person  on events for all tests, we're going to have to add the respective person_id for each `_create_event` call

File: ee/clickhouse/queries/funnels/funnel_event_query.py [hunk 4]
```diff
@@ -90,6 +89,22 @@ def get_query(self, entities=None, entity_name="events", skip_entity_filter=Fals
...
+            ]
+        else:
+            return [
+                f", {self.PERSON_TABLE_ALIAS}.{column_name} as {column_name}"
+                for column_name in self._person_query.fields
+            ]
+
+    def _person_properties_to_query(self) -> List[str]:
+        properties_to_query = self._column_optimizer._used_properties_with_type("person")
+        return [property_name for property_name, _, _ in set(properties_to_query)]
```
Comment (EDsCODE) [inline ee/clickhouse/queries/funnels/funnel_event_query.py:106]: Manually pull out person columns here that may be relevant to query

**Comment by EDsCODE** [comment]
Closing in favor of #9645

**Pull request closed without merging**